### PR TITLE
Add SingleIOBuf for efficient flatbuffers serialization/deserialization

### PR DESCRIPTION
**Pull request by Q1ngbo**
### What problem does this PR solve?

Issue Number:
#2354  #978

Problem Summary:
Hi, We plan to migrate the flatbuffers protocol used internally by Baidu to the community, which is expected to be divided into the following three steps:
1. Submit the data structure SingleIOBuf, which is used for Flatbuffers serialization and deserialization, to the apache/brpc repository;
2. Submit message construction and serialization/deserialization methods to the google/flatbuffers repository;
3. Add Flatbuffers-based RPC handling policy to apache/brpc.
These steps have a sequential dependency, and this commit completes the first step.
Flatbuffers requires that data to be serialized/deserialized be stored in a contiguous memory buffer. To meet this requirement, we designed SingleIOBuf. Unlike brpc’s existing IOBuf, SingleIOBuf uses a single Block to hold contiguous data, which better aligns with Flatbuffers' memory layout requirements. We also defined corresponding memory allocation and release interfaces to support efficient management of SingleIOBuf.

### What is changed and the side effects?

Changed:

Side effects:
- Performance effects:

- Breaking backward compatibility: 

---
### Check List:
- Please make sure your changes are compilable.
- When providing us with a new feature, it is best to add related tests.
- Please follow [Contributor Covenant Code of Conduct](https://github.com/apache/brpc/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: src/butil/iobuf.cpp [hunk 1]
```diff
@@ -1976,6 +1976,259 @@ size_t IOBufCutter::cutn(void* out, size_t n) {
     }
 }
 
+SingleIOBuf::SingleIOBuf()
```
Comment (wwbmmm) [inline src/butil/iobuf.cpp]: maybe move these code to a seperate single_iobuf.cpp and single_iobuf.h, this file is too long

File: src/butil/iobuf.h [hunk 2]
```diff
@@ -545,6 +547,47 @@ class IOBufCutter {
...
+// It is primarily used to efficiently serialize and deserialize 
+// RPC requests in flatbuffers.
+class SingleIOBuf {
+public:
+    SingleIOBuf();
+    ~SingleIOBuf();
+    SingleIOBuf(const IOBuf::BlockRef& ref);
+    SingleIOBuf(const SingleIOBuf& other);
+    SingleIOBuf& operator=(const SingleIOBuf& rhs);
+    void* alloc(uint32_t size);
```
Comment (wwbmmm) [inline src/butil/iobuf.h]: better add some comment to explain the function of each method

**Comment by zhangqiongyu** [comment]
Very useful feature for me!

File: src/butil/iobuf.h (hunk 2 above)
Comment (ivanallen) [inline src/butil/iobuf.h]: Why not use `allocate`？It will correspond to `deallocate` below.

File: src/butil/iobuf.cpp [hunk 3]
```diff
@@ -1976,6 +1976,259 @@ size_t IOBufCutter::cutn(void* out, size_t n) {
...
+        _cur_ref.length = msg_size;
+        _cur_ref.block = b;
+        b->size += msg_size;
+        _cur_ref.block->inc_ref();
+        return true;
+    }
+    return false;
+}
+
+void SingleIOBuf::append_to(IOBuf* buf) const {
```
Comment (yanglimingcn) [inline src/butil/iobuf.cpp]: In RPC scenarios, data length is dynamic, while SingleIOBuf typically manages a fixed-size memory block. would you please explain how to use it in PRC parse message and pack message？
for example, when receive a message we don't know the length of this message, how to allocate memory, if memory is short, we need to reallocate and need memcopy?

Comment (chenBright) [reply src/butil/iobuf.cpp]: I'm also curious about the implementation here.

File: src/butil/iobuf.cpp (hunk 1 above)
Comment (Q1ngbo) [reply src/butil/iobuf.cpp]: I agree that. But to do this properly, we would need to refactor iobuf.h and iobuf.cpp—for example, moving the definition of IOBuf::Block into iobuf.h and adding declarations of the butil::iobuf functions from iobuf.cpp. This would involve fairly substantial changes to these files.
Could you please confirm if such a level of modification is acceptable?

File: src/butil/iobuf.cpp (hunk 3 above)
Comment (Q1ngbo) [reply src/butil/iobuf.cpp]: 在flatbuffers中用户使用fbs文件定义message格式（类似于proto）。所以在用户应用程序中创建请求时可以按消息大小分配内存。构造出的消息内容保存在SingleIOBuf中，最终它会通过SingleIOBuf::append_to添加到IOBuf中发送出去。brpc接收到的消息同样保存在IOBuf中。因此在解析请求时，将调用SingleIOBuf::assign将IOBuf assign给SingleIOBuf, 如果IOBuf的单个Block能容纳整个message，则直接将该Block赋值给SingleIOBuf，否则会根据消息大小重新申请足够大的Block，再从传入的IOBuf中拷贝数据。

RPC message的Serialize与Deserialize过程将在flatbuffers中实现。

Comment (yanglimingcn) [reply src/butil/iobuf.cpp]: 所以接收消息这块可能会有拷贝，这块最佳实践是评估请求大小，最好是一个Block能容纳得下。

File: src/butil/iobuf.cpp (hunk 1 above)
Comment (wwbmmm) [reply src/butil/iobuf.cpp]: Yes, that's acceptable. I think you can move the definition of IOBuf::Block and butil::iobuf functions to iobuf_inl.h, not iobuf.h, because they are not user interfaces.

File: src/butil/iobuf.cpp (hunk 3 above)
Comment (Q1ngbo) [reply src/butil/iobuf.cpp]: > 所以接收消息这块可能会有拷贝，这块最佳实践是评估请求大小，最好是一个Block能容纳得下。

是的，当前单个Block的默认大小是8KB（由DEFAULT_BLOCK_SIZE指定）。

File: src/butil/iobuf.cpp (hunk 1 above)
Comment (Q1ngbo) [reply src/butil/iobuf.cpp]: > Yes, that's acceptable. I think you can move the definition of IOBuf::Block and butil::iobuf functions to iobuf_inl.h, not iobuf.h, because they are not user interfaces.

Done.

File: src/butil/iobuf.h (hunk 2 above)
Comment (Q1ngbo) [reply src/butil/iobuf.h]: > better add some comment to explain the function of each method

Done.

Comment (Q1ngbo) [reply src/butil/iobuf.h]: Done.

File: src/butil/single_iobuf.cpp [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+    }
+    if (BAIDU_UNLIKELY(!_cur_block)) {
+        errno = ENOMEM;
+        _block_size = 0;
+        return NULL;
+    }
+    return _cur_block;
+}
+
+void SingleIOBuf::memcpy_downward(void* old_p, uint32_t old_size,
```
Comment (ivanallen) [inline src/butil/single_iobuf.cpp:137]: - Could you explain what in_use_back and in_use_front are and how to use them?
- Maybe you should check `new_size >= in_use_front + in_use_back`  and `old_size >= in_use_front + in_use_back`?

File: test/iobuf_unittest.cpp [hunk 5]
```diff
@@ -1890,4 +1891,46 @@ TEST_F(IOBufTest, reserve_aligned) {
...
+    ASSERT_EQ(buf2.length(), total_len);
+    butil::SingleIOBuf sbuf2;
+    sbuf2.swap(sbuf);
+    ASSERT_EQ(sbuf.get_length(), 0);
+    ASSERT_EQ(sbuf2.get_length(), total_len);
+    sbuf2.reset();
+    ASSERT_EQ(0, sbuf2.get_length());
+    void* buf = sbuf.allocate(1024);
+    ASSERT_TRUE(NULL != buf);
+    buf = sbuf.reallocate_downward(16384, 0, 0);
```
Comment (ivanallen) [inline test/iobuf_unittest.cpp:1926]: Could you add some corner cases for reallocate_downward?

File: src/butil/single_iobuf.cpp [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    reset();
+    _block_size = 0;
+    if (rhs._cur_ref.block != NULL) {
+        _cur_ref = rhs._cur_ref;
+        _cur_ref.block->inc_ref();
+    }
+    return *this;
+}
+
+void SingleIOBuf::swap(SingleIOBuf& other) {
```
Comment (ivanallen) [inline src/butil/single_iobuf.cpp:67]: Maybe should check `this != &other` that can help you return earlier.

File: src/butil/single_iobuf.cpp [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+SingleIOBuf::~SingleIOBuf() {
+    reset();
+}
+
+SingleIOBuf& SingleIOBuf::operator=(const SingleIOBuf& rhs) {
+    reset();
+    _block_size = 0;
+    if (rhs._cur_ref.block != NULL) {
```
Comment (yanglimingcn) [inline src/butil/single_iobuf.cpp:60]: Is it reasonable to set _block_size=0 after assignment? use swap function?

File: src/butil/single_iobuf.h [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    void memcpy_downward(void* old_p, uint32_t old_size,
+                        void* new_p, uint32_t new_size,
+                        uint32_t in_use_back, uint32_t in_use_front);
+
+    // Allocates and returns a memory block large enough to hold the specified data size, 
+    // reusing an existing block when possible or creating a new one if necessary.
+    IOBuf::Block* alloc_block_by_size(uint32_t data_size);
+    
+private:
+    // Current block the SingleIOBuf used.
```
Comment (yanglimingcn) [inline src/butil/single_iobuf.h]: is this only one block in SingleIOBuf？

File: src/butil/single_iobuf.cpp [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+    }
+    return NULL;
+}
+
+uint32_t SingleIOBuf::get_length() const {
+    return _cur_ref.length;
+}
+
+void SingleIOBuf::reset() {
+    if (_cur_block) {
```
Comment (yanglimingcn) [inline src/butil/single_iobuf.cpp:189]: when _cur_block not null and _cur_ref.block is null, I think they should always consistent

File: src/butil/single_iobuf.cpp (hunk 4 above)
Comment (Q1ngbo) [reply src/butil/single_iobuf.cpp:137]: 函数memcpy_downward被reallocate_downward调用，后者是对google/flatbuffers中ReallocateDownward的适配。ReallocateDownward作用是重分配用于序列化的缓冲区，它是一维结构。序列化所使用的flatbuffers vtable元信息被存储在了缓冲区开头；写入数据时是从缓冲区末尾往前写的。因此in_use_back被赋值为了旧缓冲区中已写入数据总大小；in_use_front表示的是元信息占用的内存大小。重申请内存后需要分别拷贝这两部分。下图或许可以更直观地说明这两个变量的作用。
image

File: src/butil/single_iobuf.cpp (hunk 7 above)
Comment (Q1ngbo) [reply src/butil/single_iobuf.cpp:60]: Yes, _block_size is used to determine whether the block should be released.

File: src/butil/single_iobuf.h (hunk 8 above)
Comment (Q1ngbo) [reply src/butil/single_iobuf.h]: Yes, That's why it is called "Single"

File: src/butil/single_iobuf.cpp [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+
+void SingleIOBuf::append_to(IOBuf* buf) const {
+    if (buf && _cur_ref.block) {
+        buf->_push_back_ref(_cur_ref);
+    }
+}
+
+int SingleIOBuf::assign_user_data(void* data, size_t size, void (*deleter)(void*)) {
+    reset();
+    if (size > 0xFFFFFFFFULL - 100) {
```
Comment (ivanallen) [inline src/butil/single_iobuf.cpp:256]: What is the exact meaning of 100? Does it refer to the reserved area of in_use_back?

File: src/butil/single_iobuf.cpp [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+void SingleIOBuf::append_to(IOBuf* buf) const {
+    if (buf && _cur_ref.block) {
+        buf->_push_back_ref(_cur_ref);
+    }
+}
+
+int SingleIOBuf::assign_user_data(void* data, size_t size, void (*deleter)(void*)) {
+    reset();
```
Comment (ivanallen) [inline src/butil/single_iobuf.cpp]: This reset may have side effects. Should it be executed after the condition check?

File: src/butil/iobuf_inl.cpp [hunk 12]
```diff
@@ -0,0 +1,116 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (wwbmmm) [inline src/butil/iobuf_inl.cpp]: 1. The file name `iobuf_inl` means there are inline functions in it. It is always for header file, not cpp file. You can just move these code to `iobuf_inl.h`, and delete this `iobuf_inl.cpp` file. 
2. When you seperate the function body from class/struct definition, that function is by default not inline. This may affect performance. To make sure the function is inline, you can keep the function body in class/struct definition, or add `inline` keyword for each function.

File: src/butil/iobuf.cpp [hunk 13]
```diff
@@ -341,7 +214,7 @@ uint32_t block_size(IOBuf::Block const* b) {
     return b->size;
 }
 
-inline IOBuf::Block* create_block(const size_t block_size) {
+IOBuf::Block* create_block(const size_t block_size) {
```
Comment (wwbmmm) [inline src/butil/iobuf.cpp]: remove the inline keyword may affect performance, you can move the function body to iobuf_inl.h, since that file is for inline functions.

File: src/butil/single_iobuf.cpp [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+        return false;
+    }
+    reset();
+    if (BAIDU_UNLIKELY(msg_size == 0)) {
+        return true;
+    }
+    const IOBuf::BlockRef& ref = buf._front_ref();
+    if (ref.length >= msg_size) {
+        _cur_ref.offset = ref.offset;
+        _cur_ref.length = msg_size;
```
Comment (yanglimingcn) [inline src/butil/single_iobuf.cpp]: reset应该放到msg_size判断后面更合适些。

File: src/butil/single_iobuf.cpp (hunk 9 above)
Comment (Q1ngbo) [reply src/butil/single_iobuf.cpp:189]: In SingleIOBuf::assign(), _cur_block is set to NULL by reset(), while _cur_ref.block may be assigned the block from the input IOBuf.

File: src/butil/single_iobuf.cpp (hunk 10 above)
Comment (Q1ngbo) [reply src/butil/single_iobuf.cpp:256]: The threshold 0xFFFFFFFFULL - 100 is chosen to stay consistent with the implementation of IOBuf::append_user_data_with_meta().

File: src/butil/iobuf.cpp (hunk 13 above)
Comment (Q1ngbo) [reply src/butil/iobuf.cpp]: Done.

**Comment by wwbmmm** [comment]
LGTM

File: src/butil/single_iobuf.cpp (hunk 4 above)
Comment (yanglimingcn) [reply src/butil/single_iobuf.cpp:137]: flatbuffer的所有数据都是从后向前写的对吧。

File: src/butil/single_iobuf.cpp (hunk 9 above)
Comment (yanglimingcn) [reply src/butil/single_iobuf.cpp:189]: 我这块不是很理解，为啥_cur_ref.block都设置为IOBuf的一个block了，_cur_block为啥不同步更新呢？

**Comment by yanglimingcn** [comment]
Add Flatbuffers-based RPC handling policy to apache/brpc.
These steps have a sequential dependency, and this commit completes the first step.
Flatbuffers requires that data to be serialized/deserialized be stored in a contiguous memory buffer. To meet this requirement, we designed SingleIOBuf. Unlike brpc’s existing IOBuf, SingleIOBuf uses a single Block to hold contiguous data, which better aligns with Flatbuffers' memory layout requirements. We also defined corresponding memory allocation and release interfaces to support efficient management of SingleIOBuf.
@Q1ngbo 想问一下，这个支持完了之后，bRPC是要实现一个新的协议吗？还是可以直接用baidu_std协议？

File: src/butil/single_iobuf.cpp (hunk 4 above)
Comment (Q1ngbo) [reply src/butil/single_iobuf.cpp:137]: > flatbuffer的所有数据都是从后向前写的对吧。

是的。

File: src/butil/single_iobuf.cpp (hunk 9 above)
Comment (Q1ngbo) [reply src/butil/single_iobuf.cpp:189]: _cur_block表示的是SingleIOBuf 自己负责分配/释放的block。_cur_ref.block可以认为是当前实际存储数据的block。
在当前reset()实现中，如果在assign里同步更新_cur_block，会导致在后续reset中错误地dec_ref同一块block两次。

**Comment by Q1ngbo** [comment]
> Add Flatbuffers-based RPC handling policy to apache/brpc. These steps have a sequential dependency, and this commit completes the first step. Flatbuffers requires that data to be serialized/deserialized be stored in a contiguous memory buffer. To meet this requirement, we designed SingleIOBuf. Unlike brpc’s existing IOBuf, SingleIOBuf uses a single Block to hold contiguous data, which better aligns with Flatbuffers' memory layout requirements. We also defined corresponding memory allocation and release interfaces to support efficient management of SingleIOBuf. @Q1ngbo 想问一下，这个支持完了之后，bRPC是要实现一个新的协议吗？还是可以直接用baidu_std协议？

是的，会实现一个新的协议。

**Review by ivanallen — APPROVED** [review]

File: src/butil/single_iobuf.cpp [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+        _cur_ref.block = ref.block;
+        _cur_ref.block->inc_ref();
+        return true;
+    } else {
+        IOBuf::Block* b = alloc_block_by_size(msg_size);
+        if (!b) {
+            return false;
```
Comment (chenBright) [inline src/butil/single_iobuf.cpp:227]: Is it better to ensure that the block is allocated successfully before reset?

File: src/butil/single_iobuf.cpp [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+    }
+}
+
+void SingleIOBuf::append_to(IOBuf* buf) const {
+    if (buf && _cur_ref.block) {
+        buf->_push_back_ref(_cur_ref);
+    }
+}
+
+int SingleIOBuf::assign_user_data(void* data, size_t size, void (*deleter)(void*)) {
```
Comment (chenBright) [inline src/butil/single_iobuf.cpp]: Like IOBuf, the deleter type should be std::function, not a function pointer.

https://github.com/apache/brpc/blob/f949962e141fecba92bbd399d1da75fe2ac6862d/src/butil/iobuf.h#L250

File: src/butil/single_iobuf.h [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    // Allocates a contiguous memory region of the specified size.
+    // Returns a pointer to the start of allocated space within the block.
+    void* allocate(uint32_t size);
+    // Deallocates the block if the given pointer matches its starting address.
+    void deallocate(void* p);
+
+    // Reallocates the current buffer to a larger size.
+    void* reallocate_downward(uint32_t new_size,
+                            uint32_t in_use_back,
+                            uint32_t in_use_front);
```
Comment (chenBright) [inline src/butil/single_iobuf.h:53]: Add a comment to explain in_use_back and in_use_front.

Comment (Q1ngbo) [reply src/butil/single_iobuf.h:53]: All done.

File: src/butil/single_iobuf.cpp (hunk 15 above)
Comment (chenBright) [reply src/butil/single_iobuf.cpp:227]: @Q1ngbo

Comment (Q1ngbo) [reply src/butil/single_iobuf.cpp:227]: Yes, it's meaningful. The latest commit has included this.

**Review by chenBright — APPROVED** [review]
LGTM

File: src/butil/single_iobuf.cpp (hunk 9 above)
Comment (yanglimingcn) [reply src/butil/single_iobuf.cpp:189]: 我理解assign和reset这俩动作肯定是要互斥执行的，能否让他们执行完之后，都达到一个最终状态呢？即，_cur_block和_cur_ref一致了。否则如果reset执行完了，这两个不一致了，理解起来有点费劲。reset其实也是可以作为一个API单独调用的。

Comment (Q1ngbo) [reply src/butil/single_iobuf.cpp:189]: _cur_block和_cur_ref不一致是在设计上有意为之的。_cur_ref.block始终指向有效数据所在的 IOBuf::Block；_cur_block是一个内存分配缓存，只用于分配内存。assign()在flatbuffers中的作用是保存接收到的message，该block大小已经固定了，所以不应该将其赋值给_cur_block。

Comment (yanglimingcn) [reply src/butil/single_iobuf.cpp:189]: 收到，理解了

**Comment by yanglimingcn** [comment]
LGTM

**Comment by zchuango** [comment]
@Q1ngbo 挺感兴趣brpc引入flatbuffers特性的，这块后续的计划是什么？目前还没看到step2和step3的pr

**Comment by AnDiXL** [comment]
> @Q1ngbo 挺感兴趣brpc引入flatbuffers特性的，这块后续的计划是什么？目前还没看到step2和step3的pr

现在有啦
